### PR TITLE
Fix scaling in Virtual XFB.

### PR DESCRIPTION
**Pull request by ZephyrSurfer**
This should make Ty the Tasmanian Tiger 3 stretch across the whole
screen in Virtual XFB. There are other games with this same issue.

See <a href="https://code.google.com/p/dolphin-emu/issues/detail?id=6750">issue 6750</a> for details.

### CODE REVIEW

**Comment by Armada651** [comment]
Looks like I forgot about that line, thanks for taking care of it. LGTM

**Comment by Armada651** [comment]
@dolphin-emu-bot rebuild

**Comment by Linktothepast** [comment]
I am having an issue with this in that it now works with virtual xfb but only with 1x ir which kinda defeats the purpose. Same was and is the case with opengl, it worked fine with virtual xfb but only when 1x ir was used.

**Comment by ZephyrSurfer** [comment]
@Linktothepast Thank you for the report. We just have to multiply the last line by the scaling factor.

I'll write it in a few minutes.

**Comment by ZephyrSurfer** [comment]
Is there any clean way to find IR scale or will I have to write in RenderBase?

**Comment by Armada651** [comment]
You can use `Renderer::EFBToScaledX` and `Renderer::EFBToScaledY` to scale horizontal or vertical values to the IR scale.

**Comment by ZephyrSurfer** [comment]
Thanks @Armada651  
I'll push another commit in a minute.

**Comment by ZephyrSurfer** [comment]
I pushed that commit.

Maybe you could try that now @Linktothepast

**Comment by Linktothepast** [comment]
I can't you are not auto trusted and i had build issues for a while and removed visual studio not building anymore.
Now let's see if this works otherwise you should ask somebody else to do it...
@dolphin-emu-bot rebuild

**Comment by ZephyrSurfer** [comment]
@Linktothepast That worked :D

**Comment by Linktothepast** [comment]
Yep, seems to fix all the games mentioned in the issue 6750 with all backends, tomorrow i will change their gameinis.

**Comment by Armada651** [comment]
@Linktothepast Have you checked both backends? I find it a little bit strange that this is required, it may be an error I introduced in the D3D backend.

**Comment by ZephyrSurfer** [comment]
@Armada651 It's existed forever basically; I think.

**Comment by ZephyrSurfer** [comment]
I tested another game this fixes the scaling with it. The game is <a href="https://wiki.dolphin-emu.org/index.php?title=Nickelodeon_Teenage_Mutant_Ninja_Turtles">Teenage Mutant Ninja Turtles</a>

**Comment by Linktothepast** [comment]
Found a d3d specific texture corruption issue due to this pr with Teenage Mutant Ninja Turtles you wrote above, text boxes ingame are black even with proper settings when using virtual xfb and they crash giving a texture corruption error with real xfb, opengl is fine. I don't think it is this pr's fault since even real xfb is affected and only d3d, just mentioning it since i discovered it due to this.

Virtual xfb, d3d: http://prntscr.com/5oavqa
Real xfb, d3d: http://prntscr.com/5oavwx

**Comment by ZephyrSurfer** [comment]
@Linktothepast  Not this PRs fault but thanks for bringing this up and since I like TMNT things I'd like to fix it.  
Someone else on the issue tracker mentioned something similar to me on <a href="https://code.google.com/p/dolphin-emu/issues/detail?id=6749">Issue 6749</a>  
So it's likely an nVidia bug/regression, could you bisect?

**Comment by Linktothepast** [comment]
It seems to have started somewhere between 4.0.2 and 4.0 -1995. I will track it down better later.

**Comment by Linktothepast** [comment]
Ok, found it, it is:
 https://el.dolphin-emu.org/download/dev/edbf90f9d25fdf5d2c080b1854695a9d77cd2134/   or   
https://github.com/dolphin-emu/dolphin/pull/336
the commit from which virtual xfb in d3d starts showing this texture corruption.

**Comment by ZephyrSurfer** [comment]
@Linktothepast Excellent, I'll try to fix this in a separate PR in a little while.

But I had meant to bisect about the crashing with Real XFB; Sorry, for not being clear.
I only have an AMD card so I cannot do it.

**Comment by Linktothepast** [comment]
Sorry, i didn't see your edit, yes both issues started with that pr with d3d. But the difference is that real xfb showed a garbled mess before that pr and simply crashed afterwards (it didn't work ok beforehand either, afterwards though it didn't work at all).

**Comment by ZephyrSurfer** [comment]
Ok so there is another virtual XFB issue that exists and it only now occured to me that logically it would also be contained in this PR that touches Virtual XFB.

TL;DR: Don't merge this yet

**Comment by ZephyrSurfer** [comment]
There is boxes of text missing in Teenage Mutant Ninja Turtles(2013) because of <a href="https://github.com/magumagu/dolphin/commit/2db2f3dce9df8aa28a7f102acb38563078bf5daa">commit 1</a> in #336  
The line in question is:
<code>u32 fbWidth = m_HorizontalStepping.FbSteps * (field == FIELD_PROGRESSIVE ? 16 : 8);</code>

Can you give this a look @magumagu

**Comment by magumagu** [comment]
What's the issue?  "boxes of text missing" isn't much of a description.

**Comment by ZephyrSurfer** [comment]
@magumagu Linktothepast already uploaded an image above.

**Comment by magumagu** [comment]
You're saying that 2db2f3d breaks Teenage Mutant Ninja Turtles(2013) with XFB enabled specifically in D3D mode (but not OpenGL)?  Probably an issue with the coordinate math in the D3D-specific XFB display code.  The PanicAlert is probably easier to track down; try setting a breakpoint in D3DTexture to figure out what's wrong with the input.

**Comment by ZephyrSurfer** [comment]
I'm not sure yet about the crash with realXFB because I can't reproduce it with AMD.
But that's where the text box stops getting rendered well in D3D Virtual XFB.
Can I give you a fifo?

**Comment by magumagu** [comment]
I'd be happy to look at it, but IIRC FIFOs don't work for XFB issues...

Might be easier to continue this discussion on IRC.

**Comment by ZephyrSurfer** [comment]
Oh well I'll track it down like you suggested.  
I think I have turn in now though.
So i'll post whatever I find tomorrow.

**Comment by ZephyrSurfer** [comment]
@Linktothepast Can you try these two builds(the first two) from PR 336 I've compiled them:
Here is a link on my <a href="https://www.dropbox.com/sh/x98lalq786obr69/AAC6ChSibfshd495gXrIlIzQa?dl=0">Dropbox</a>

**Comment by Linktothepast** [comment]
They both have both issues, but i would like to be more specific about the real xfb crash that i first get a black screen in dolphin and a driver crash (your gpu stopped responding... etc..)

**Comment by ZephyrSurfer** [comment]
@Linktothepast That just means the first commit introduced the crashing. So whenever I find what's wrong with the D3D math it'll fix both issues.

**Comment by degasus** [comment]
What is the status of this PR? It's still marked as WIP, but imo it's likely finished.

**Comment by ZephyrSurfer** [comment]
@degasus Yeah finished. I'm going to leave other XFB issues to a different PR.
